README.md
# Junction coverage compatibility scores

This repository contains an R package aimed at facilitating the calculation of gene-wise junction coverage compatibility (JCC) scores, as described in 

- Soneson C, Love MI, Patro R, Hussain S, Malhotra D and Robinson MD: A junction coverage compatibility score to quantify the reliability of transcript abundance estimates and annotation catalogs. [bioRxiv doi:10.1101/378539](https://www.biorxiv.org/content/early/2018/07/28/378539).

The JCC scores are aimed at detecting genes for which estimated transcript abundances are unreliable, either because of problems in the transcript abundance estimation or because of missing or wrongly annotated reference transcripts. It does so by comparing the number of reads predicted to align across each junction, inferred from the transcript abundances and a fragment bias model, to the observed number of junction-spanning reads, obtained via alignment of the reads to the genome. A high JCC score for a gene indicates that the estimated abundances for the corresponding transcripts are unreliable and should be treated with care in downstream analyses. 

## Installation

The `jcc` package can be installed using `devtools`:

```
library(devtools)
install_github("csoneson/jcc")
```

or the `BiocManager` CRAN package

```
BiocManager::install("csoneson/jcc")
```


NEWS
CHANGES IN VERSION 0.1.0
------------------------
  o Initial version of the jcc package

R/combineCoverages.R
#' Combine observed and predicted coverages
#'
#' This function combines predicted junction coverages (from
#' \code{\link{scaleTxCoverages}}) and observed junction coverages (obtained
#' from aligning the reads to the genome and counting the number of reads
#' spanning each junction). It also summarized transcript abundances on the gene
#' level and adds information about the number/fraction of uniquely mapping
#' reads and multimapping reads spanning each junction.
#'
#' @param junctionCounts A \code{data.frame} with observed junction coverages.
#'   Should have columns \code{seqnames}, \code{start}, \code{end},
#'   \code{strand}, \code{uniqreads} and \code{mmreads}.
#' @param junctionPredCovs A \code{data.frame} with predicted junction
#'   coverages.
#' @param txQuants A \code{data.frame} with estimated transcript abundances.
#'
#' @return A list with three elements: \describe{ \item{\code{junctionCovs}:}{A
#'   \code{GRanges} object with predicted and observed junction coverages.}
#'   \item{\code{txQuants}:}{A \code{tibble} with estimated transcript
#'   abundances.} \item{\code{geneQuants}:}{A \code{tibble} with summarized gene
#'   abundances.} }
#'
#' @author Charlotte Soneson
#'
#' @export
#'
#' @references
#' Soneson C, Love MI, Patro R, Hussain S, Malhotra D, Robinson MD:
#' A junction coverage compatibility score to quantify the reliability of
#' transcript abundance estimates and annotation catalogs. bioRxiv
#' doi:10.1101/378539 (2018)
#'
#' @examples
#' \dontrun{
#' gtf <- system.file("extdata/Homo_sapiens.GRCh38.90.chr22.gtf.gz",
#'                    package = "jcc")
#' bam <- system.file("extdata/reads.chr22.bam", package = "jcc")
#' biasMod <- fitAlpineBiasModel(gtf = gtf, bam = bam,
#'                               organism = "Homo_sapiens",
#'                               genome = Hsapiens, genomeVersion = "GRCh38",
#'                               version = 90, minLength = 230,
#'                               maxLength = 7000, minCount = 10,
#'                               maxCount = 10000, subsample = TRUE,
#'                               nbrSubsample = 30, seed = 1, minSize = NULL,
#'                               maxSize = 220, verbose = TRUE)
#' tx2gene <- readRDS(system.file("extdata/tx2gene.sub.rds", package = "jcc"))
#' predCovProfiles <- predictTxCoverage(biasModel = biasMod$biasModel,
#'                                      exonsByTx = biasMod$exonsByTx,
#'                                      bam = bam, tx2gene = tx2gene,
#'                                      genome = Hsapiens,
#'                                      genes = c("ENSG00000070371",
#'                                                "ENSG00000093010"),
#'                                      nCores = 1, verbose = TRUE)
#' txQuants <- readRDS(system.file("extdata/quant.sub.rds", package = "jcc"))
#' txsc <- scaleTxCoverages(txCoverageProfiles = predCovProfiles,
#'                          txQuants = txQuants, tx2gene = tx2gene,
#'                          strandSpecific = TRUE, methodName = "Salmon",
#'                          verbose = TRUE)
#' jcov <- read.delim(system.file("extdata/sub.SJ.out.tab", package = "jcc"),
#'                    header = FALSE, as.is = TRUE) %>%
#'  setNames(c("seqnames", "start", "end", "strand", "motif", "annot",
#'            "uniqreads", "mmreads", "maxoverhang")) %>%
#'  dplyr::mutate(strand = replace(strand, strand == 1, "+")) %>%
#'  dplyr::mutate(strand = replace(strand, strand == 2, "-")) %>%
#'  dplyr::select(seqnames, start, end, strand, uniqreads, mmreads) %>%
#'  dplyr::mutate(seqnames = as.character(seqnames))
#' combCov <- combineCoverages(junctionCounts = jcov,
#'                             junctionPredCovs = txsc$junctionPredCovs,
#'                             txQuants = txsc$txQuants)
#' }
#'
#' @importFrom dplyr summarize mutate select ungroup group_by left_join distinct
#'   arrange filter everything
#'
combineCoverages <- function(junctionCounts, junctionPredCovs,
                             txQuants) {

  ## Process observed junction counts
  jcovnostrand <- junctionCounts %>%
    dplyr::group_by(seqnames, start, end) %>%
    dplyr::summarize(uniqreads = sum(uniqreads),
                     mmreads = sum(mmreads)) %>%
    dplyr::mutate(strand = "*") %>% dplyr::ungroup() %>%
    dplyr::select(seqnames, start, end, strand, uniqreads, mmreads) %>%
    as.data.frame()
  jcov <- rbind(junctionCounts, jcovnostrand)

  ## Process predicted junction counts
  junctionPredCovs <- junctionPredCovs %>%
    dplyr::group_by(seqnames, start, end, gene) %>%
    dplyr::mutate(transcript = paste(unique(strsplit(
      paste(unique(transcript), collapse = ","),
      ",")[[1]]), collapse = ",")) %>% ## to make sure we have the same
                                       ## transcript combination for a given
                                       ## junction across all methods
    dplyr::ungroup() %>%
    dplyr::mutate(predCov = replace(predCov, is.na(predCov), 0)) %>%
    dplyr::left_join(jcov, by = c("seqnames", "start", "end", "strand")) %>%
    dplyr::mutate(uniqreads = replace(uniqreads, is.na(uniqreads), 0),
                  mmreads = replace(mmreads, is.na(mmreads), 0)) %>%
    dplyr::mutate(fracunique = uniqreads/(uniqreads + mmreads)) %>%
    dplyr::mutate(fracunique = replace(fracunique, is.na(fracunique), 1)) %>%
    dplyr::ungroup()

  ## Assign a junction ID to each junction
  j0 <- junctionPredCovs %>% dplyr::select(seqnames, start, end, gene) %>%
    dplyr::distinct() %>% dplyr::group_by(gene) %>% dplyr::arrange(start) %>%
    dplyr::mutate(junctionid = paste0("J", seq_len(length(start)))) %>%
    dplyr::ungroup()

  junctionPredCovs <- junctionPredCovs %>% dplyr::left_join(j0) %>%
    dplyr::select(junctionid, dplyr::everything(), transcript)

  geneQuants <- txQuants %>% dplyr::group_by(gene, method) %>%
    dplyr::mutate(TPMrel = TPM/sum(TPM)) %>%
    dplyr::mutate(TPMrel = replace(TPMrel, is.na(TPMrel), 0)) %>%
    dplyr::summarize(count = sum(count),
                     TPM = sum(TPM),
                     nbr_expressed_transcripts = sum(TPM > 0),
                     nbr_expressed_transcripts_5p = sum(TPMrel > 0.05),
                     covOKfraction = length(which(covNote == "covOK"))/
                       length(covNote)) %>%
    dplyr::ungroup()

  totjunctionreads <- junctionPredCovs %>%
    dplyr::filter(method == method[1]) %>%
    dplyr::select(gene, uniqreads, mmreads) %>%
    dplyr::group_by(gene) %>% dplyr::summarize(uniqjuncreads = sum(uniqreads),
                                               mmjuncreads = sum(mmreads))
  geneQuants <- dplyr::left_join(geneQuants, totjunctionreads, by = "gene") %>%
    dplyr::mutate(uniqjuncreads = replace(uniqjuncreads,
                                          is.na(uniqjuncreads), 0)) %>%
    dplyr::mutate(mmjuncreads = replace(mmjuncreads,
                                        is.na(mmjuncreads), 0)) %>%
    dplyr::mutate(uniqjuncfraction =
                    uniqjuncreads/(uniqjuncreads + mmjuncreads)) %>%
    dplyr::mutate(uniqjuncfraction = replace(uniqjuncfraction,
                                             is.na(uniqjuncfraction), 1))

  ## Add number of junctions per gene (from the junction summary). If a gene is
  ## not in the junction summary, it means that it doesn't have any annotated
  ## junctions.
  geneQuants <- geneQuants %>%
    dplyr::left_join(junctionPredCovs %>%
                       dplyr::select(gene, method) %>%
                       dplyr::distinct(),
                     by = c("gene", "method"))

  list(junctionCovs = as.data.frame(junctionPredCovs),
       txQuants = as.data.frame(txQuants),
       geneQuants = as.data.frame(geneQuants))
}


R/scaleTxCoverages.R
#' Scale transcript coverages by abundance estimate
#'
#' This function takes a list of predicted transcript coverage profiles (from
#' \code{\link{predictTxCoverage}}) and corresponding transcript abundance
#' estimates (from, e.g., Salmon, kallisto, StringTie, RSEM, or any other
#' transcript abundance estimation software), and scales the coverage profiles
#' by the abundance estimates in order to determine the predicted number of
#' reads covering each position in the transcript. In addition, it extracts the
#' predicted number of reads spanning each junction in each transcript, and sums
#' up these for each annotated junction, across all transcripts containing the
#' junction.
#'
#' @param txCoverageProfiles A \code{list} of predicted transcript coverage
#'   profiles (output from \code{\link{predictTxCoverage}}).
#' @param txQuants A \code{data.frame} with estimated transcript abundances.
#'   Must contain columns named \code{transcript}, \code{TPM} and \code{count},
#'   all other columns will be ignored.
#' @param tx2gene A \code{data.frame} with transcript-to-gene mapping. Must have
#'   at least two columns, named \code{tx} and \code{gene}.
#' @param strandSpecific Logical, whether or not the data is strand-specific.
#'   Determines whether or not the strand of a junction should be taken into
#'   account when the coverages are summed up across transcripts.
#' @param methodName Name to use for the abundance estimation method in the
#'   output table.
#' @param verbose Logical, whether to print progress messages.
#'
#' @return A list with two elements:
#' \describe{
#' \item{\code{junctionPredCovs}:}{A \code{tibble} with the predicted coverage
#' for each junction, scaled by the estimated transcript abundances and
#' summarized across all transcripts.}
#' \item{\code{txQuants}:}{A \code{tibble} with estimated transcript abundances,
#' including information about whether or not the coverage profile could be
#' predicted ('covOK') or if a uniform coverage was imposed ('covError' or
#' 'covNA').}
#' }
#'
#' @author Charlotte Soneson
#'
#' @export
#'
#' @references
#' Soneson C, Love MI, Patro R, Hussain S, Malhotra D, Robinson MD: A junction
#' coverage compatibility score to quantify the reliability of transcript
#' abundance estimates and annotation catalogs. bioRxiv doi:10.1101/378539
#' (2018)
#'
#' @examples
#' \dontrun{
#' gtf <- system.file("extdata/Homo_sapiens.GRCh38.90.chr22.gtf.gz",
#'                    package = "jcc")
#' bam <- system.file("extdata/reads.chr22.bam", package = "jcc")
#' biasMod <- fitAlpineBiasModel(gtf = gtf, bam = bam,
#'                               organism = "Homo_sapiens",
#'                               genome = Hsapiens, genomeVersion = "GRCh38",
#'                               version = 90, minLength = 230,
#'                               maxLength = 7000, minCount = 10,
#'                               maxCount = 10000, subsample = TRUE,
#'                               nbrSubsample = 30, seed = 1, minSize = NULL,
#'                               maxSize = 220, verbose = TRUE)
#' tx2gene <- readRDS(system.file("extdata/tx2gene.sub.rds", package = "jcc"))
#' predCovProfiles <- predictTxCoverage(biasModel = biasMod$biasModel,
#'                                      exonsByTx = biasMod$exonsByTx,
#'                                      bam = bam, tx2gene = tx2gene,
#'                                      genome = Hsapiens,
#'                                      genes = c("ENSG00000070371",
#'                                                "ENSG00000093010"),
#'                                      nCores = 1, verbose = TRUE)
#' txQuants <- readRDS(system.file("extdata/quant.sub.rds", package = "jcc"))
#' txsc <- scaleTxCoverages(txCoverageProfiles = predCovProfiles,
#'                          txQuants = txQuants, tx2gene = tx2gene,
#'                          strandSpecific = TRUE, methodName = "Salmon",
#'                          verbose = TRUE)
#' }
#'
#' @importFrom dplyr mutate group_by ungroup left_join select distinct select
#'   rename
#'
scaleTxCoverages <- function(txCoverageProfiles, txQuants, tx2gene,
                             strandSpecific, methodName, verbose = FALSE) {

  ## Keep only transcript ID, count and TPM columns in quantification table
  txQuants <- txQuants %>% dplyr::select(transcript, count, TPM)

  transcripts <- names(txCoverageProfiles)
  names(transcripts) <- transcripts

  ## Go through all transcripts and scale predicted junction coverage by the
  ## estimated abundance
  scaledcovs <- lapply(transcripts, function(tx) {
    tryCatch({
      ab <- txQuants$count[txQuants$transcript == tx]
      ## if the transcript is not present in the quantification file
      if (length(ab) == 0 || is.na(ab)) ab <- 0
      m <- txCoverageProfiles[[tx]]$junctionCov
      m$pred.cov <- m$pred.cov / max(1e-10,
                                     sum(txCoverageProfiles[[tx]]$predCov)) *
        ab * txCoverageProfiles[[tx]]$aveFragLength
      as.data.frame(m) %>% dplyr::mutate(transcript = tx)
    }, error = function(e) NULL)
  })

  ## Combine estimates for all transcripts
  junctionPredCovs <- do.call(rbind, scaledcovs)
  if (!strandSpecific) {
    junctionPredCovs$strand <- "*"
  }

  ## Sum coverages of the same junction from different transcripts. Each
  ## junction is present once per gene it is included in.
  junctionPredCovs <- junctionPredCovs %>%
    dplyr::group_by(seqnames, start, end, strand) %>%
    dplyr::mutate(pred.cov = sum(pred.cov)) %>%
    dplyr::rename(predCov = pred.cov) %>%
    dplyr::ungroup() %>%
    dplyr::left_join(tx2gene %>% dplyr::select(tx, gene),
                     by = c("transcript" = "tx")) %>%
    dplyr::group_by(seqnames, start, end, strand, gene) %>%
    dplyr::mutate(transcript = paste(transcript, collapse = ",")) %>%
    dplyr::ungroup() %>%
    dplyr::distinct() %>%
    dplyr::mutate(method = methodName)

  ## Get coverage "note" for each transcript
  covNotes <- vapply(transcripts, function(tx) {
    txCoverageProfiles[[tx]]$note
  }, character(1))

  ## Add gene info to transcript quant table
  txQuants <- txQuants %>% dplyr::left_join(tx2gene %>% dplyr::select(tx, gene),
                                        by = c("transcript" = "tx")) %>%
    dplyr::mutate(method = methodName) %>%
    dplyr::inner_join(data.frame(transcript = names(covNotes),
                                 covNote = covNotes,
                                 stringsAsFactors = FALSE),
                      by = "transcript") %>%
    dplyr::select(transcript, gene, count, TPM, method, covNote)

  list(junctionPredCovs = as.data.frame(junctionPredCovs),
       txQuants = as.data.frame(txQuants))
}


R/predictTxCoverage.R
#' Predict coverage profiles of transcripts
#'
#' This function predicts coverage profiles for each transcript in a specified
#' set of genes (or all transcripts in the reference catalog), based on the
#' fragment bias model estimated by \code{alpine}.
#'
#' @param biasModel Bias model from \code{alpine}. Can be generated using the
#'   \code{\link{fitAlpineBiasModel}} function.
#' @param exonsByTx A \code{GRangesList} grouping exons by transcript. Can be
#'   generated using the \code{\link{fitAlpineBiasModel}} function.
#' @param bam Path to bam file with read alignments to the genome.
#' @param genes Character vector of gene IDs. Coverage profiles will be
#'   estimated for all transcripts annotated to any of these genes. If set to
#'   \code{NULL}, coverage profiles will be estimated for all transcripts in the
#'   reference catalog.
#' @param nCores Integer, number of cores to use for parallel computations.
#' @param tx2gene A \code{data.frame} with transcript-to-gene mapping. Must have
#'   at least two columns, named \code{tx} and \code{gene}.
#' @param genome A \code{BSgenome} object.
#' @param verbose Logical, whether to print progress messages.
#'
#' @return A list with one element per transcript. Each element is a list with
#'   five components:
#' \describe{
#' \item{\code{predCov}:}{The predicted coverage profile for the transcript.}
#' \item{\code{strand}:}{The annotated strand for the transcript.}
#' \item{\code{junctionCov}:}{A \code{data.frame} with genomic coordinates and
#' predicted read coverages for each junction in the transcript.}
#' \item{\code{aveFragLength}:}{The estimated average fragment length.}
#' \item{\code{note}:}{Either 'covOK', 'covError' or 'covNA'. If 'covOK',
#' \code{alpine} was able to predict a coverage profile. If 'covError' or
#' 'covNA', the coverage profile could not be predicted, most likely because the
#' transcript is shorter than the fragment length or because no reads in the BAM
#' file overlapped the transcript. In both these cases, we impose a uniform
#' coverage profile for the transcript.}
#' }
#'
#' @author Charlotte Soneson, Michael I Love
#'
#' @export
#'
#' @references
#' Soneson C, Love MI, Patro R, Hussain S, Malhotra D, Robinson MD: A junction
#' coverage compatibility score to quantify the reliability of transcript
#' abundance estimates and annotation catalogs. bioRxiv doi:10.1101/378539
#' (2018)
#'
#' Love MI, Hogenesch JB, Irizarry RA: Modeling of RNA-seq fragment sequence
#' bias reduces systematic errors in transcript abundance estimation. Nature
#' Biotechnology 34(12):1287-1291 (2016).
#'
#' @examples
#' \dontrun{
#' gtf <- system.file("extdata/Homo_sapiens.GRCh38.90.chr22.gtf.gz",
#'                    package = "jcc")
#' bam <- system.file("extdata/reads.chr22.bam", package = "jcc")
#' biasMod <- fitAlpineBiasModel(gtf = gtf, bam = bam,
#'                               organism = "Homo_sapiens",
#'                               genome = Hsapiens, genomeVersion = "GRCh38",
#'                               version = 90, minLength = 230,
#'                               maxLength = 7000, minCount = 10,
#'                               maxCount = 10000, subsample = TRUE,
#'                               nbrSubsample = 30, seed = 1, minSize = NULL,
#'                               maxSize = 220, verbose = TRUE)
#' tx2gene <- readRDS(system.file("extdata/tx2gene.sub.rds", package = "jcc"))
#' predCovProfiles <- predictTxCoverage(biasModel = biasMod$biasModel,
#'                                      exonsByTx = biasMod$exonsByTx,
#'                                      bam = bam, tx2gene = tx2gene,
#'                                      genome = Hsapiens,
#'                                      genes = c("ENSG00000070371",
#'                                                "ENSG00000093010"),
#'                                      nCores = 1, verbose = TRUE)
#' }
#'
#' @importFrom alpine predictCoverage
#' @importFrom GenomicRanges setdiff mcols
#' @importFrom S4Vectors Rle
#' @importFrom parallel mclapply
#'
predictTxCoverage <- function(biasModel, exonsByTx, bam, genes, nCores,
                              tx2gene, genome, verbose = FALSE) {

  ## Estimate average fragment length
  aveFragLength <- sum(biasModel$`1`$fraglen.density$x *
                         biasModel$`1`$fraglen.density$y/
                         sum(biasModel$`1`$fraglen.density$y))

  ## Load bam file
  bamFiles <- bam
  names(bamFiles) <- "1"

  ## Get all transcripts
  transcripts <- names(exonsByTx)
  if (!is.null(genes)) {
    transcriptsInProvidedGenes <- tx2gene$tx[tx2gene$gene %in% genes]
    transcripts <- intersect(transcripts, transcriptsInProvidedGenes)
  }
  names(transcripts) <- transcripts
  if (verbose) message(paste0("Predicting coverage for ",
                              length(transcripts), " transcripts..."))

  res <- parallel::mclapply(transcripts, function(tx) {
    if (verbose) message(tx)
    ## Get transcript model
    txmod <- exonsByTx[[tx]]
    txmods <- sort(exonsByTx[[tx]])

    ## Predict coverage
    m <- tryCatch({
      a <- alpine::predictCoverage(gene = txmod,
                                   bam.files = bamFiles,
                                   fitpar = biasModel,
                                   genome = genome,
                                   model.names = "all")
      if (!is.null(a$`1`$pred.cov$all)) {
        list(covr = a$`1`$pred.cov$all,
             note = "covOK")
      } else {
        ## E.g. if there are no reads mapping to the gene locus
        ## Assume uniform coverage
        list(covr = S4Vectors::Rle(rep(1, sum(width(txmod)) - 1)),
             note = "covNA")
      }
    },
    error = function(e) {
      print(paste(c(tx, e), collapse = ": "))
      ## E.g. if the transcript is shorter than the fragment length
      ## Assume uniform coverage
      list(covr = S4Vectors::Rle(rep(1, sum(width(txmod)) - 1)),
           note = "covError")
    })

    note <- m$note
    m <- m$covr

    ## Get predicted coverage for junctions
    if (verbose) message("Predicting junction coverages...")
    junctions <- GenomicRanges::setdiff(range(txmods), txmods)
    if (all(strand(txmods) == "+")) {
      junctionpos <- cumsum(width(txmods))
      junctionpos <- junctionpos[-length(junctionpos)]
      GenomicRanges::mcols(junctions)$pred.cov <- as.numeric(m)[junctionpos]
      strand <- "+"
    } else if (all(strand(txmods) == "-")) {
      junctionpos <- cumsum(width(rev(txmods)))
      junctionpos <- junctionpos[-length(junctionpos)]
      GenomicRanges::mcols(junctions)$pred.cov <-
        rev(as.numeric(m)[junctionpos])
      strand <- "-"
    } else {
      strand <- "mixed"
    }

    list(predCov = m, strand = strand, junctionCov = junctions,
         aveFragLength = aveFragLength, note = note)

  }, mc.preschedule = FALSE, mc.cores = nCores)

  res
}


R/plotGeneSummary.R
#' Plot JCC score summary for a given gene
#'
#' This function generates a multi-panel plot, including (i) the observed
#' coverage and gene model, (ii) the estimated transcript abundances and (iii)
#' the observed and predicted junction coverages within the gene.
#'
#' @param gtf Path to gtf file with genomic features. Preferably in Ensembl
#'   format.
#' @param junctionCovs A \code{data.frame} with junction coverage information,
#'   output from \code{\link{calculateJCCScores}}.
#' @param useGene A character string giving the name of the gene to plot.
#' @param bwFile Path to a bigwig file generated from a genome alignment BAM
#'   file.
#' @param txQuants A \code{data.frame} with estimated transcript abundances,
#'   output from \code{\link{combineCoverages}}.
#' @param txCol The column in the gtf file that contains the transcript ID.
#' @param geneCol The column in the gtf file that contains the gene ID.
#' @param exonCol The column in the gtf file that contains the exon ID.
#'
#' @return A \code{ggplot} object with a multi-panel plot. The size is optimized
#'   for display in a device of size approximately 12in wide and 10in high.
#'
#' @author Charlotte Soneson
#'
#' @export
#'
#' @references
#' Soneson C, Love MI, Patro R, Hussain S, Malhotra D, Robinson MD: A junction
#' coverage compatibility score to quantify the reliability of transcript
#' abundance estimates and annotation catalogs. bioRxiv doi:10.1101/378539
#' (2018)
#'
#' @examples
#' \dontrun{
#' gtf <- system.file("extdata/Homo_sapiens.GRCh38.90.chr22.gtf.gz",
#'                    package = "jcc")
#' bam <- system.file("extdata/reads.chr22.bam", package = "jcc")
#' biasMod <- fitAlpineBiasModel(gtf = gtf, bam = bam,
#'                               organism = "Homo_sapiens",
#'                               genome = Hsapiens, genomeVersion = "GRCh38",
#'                               version = 90, minLength = 230,
#'                               maxLength = 7000, minCount = 10,
#'                               maxCount = 10000, subsample = TRUE,
#'                               nbrSubsample = 30, seed = 1, minSize = NULL,
#'                               maxSize = 220, verbose = TRUE)
#' tx2gene <- readRDS(system.file("extdata/tx2gene.sub.rds", package = "jcc"))
#' predCovProfiles <- predictTxCoverage(biasModel = biasMod$biasModel,
#'                                      exonsByTx = biasMod$exonsByTx,
#'                                      bam = bam, tx2gene = tx2gene,
#'                                      genome = Hsapiens,
#'                                      genes = c("ENSG00000070371",
#'                                                "ENSG00000093010"),
#'                                      nCores = 1, verbose = TRUE)
#' txQuants <- readRDS(system.file("extdata/quant.sub.rds", package = "jcc"))
#' txsc <- scaleTxCoverages(txCoverageProfiles = predCovProfiles,
#'                          txQuants = txQuants, tx2gene = tx2gene,
#'                          strandSpecific = TRUE, methodName = "Salmon",
#'                          verbose = TRUE)
#' jcov <- read.delim(system.file("extdata/sub.SJ.out.tab", package = "jcc"),
#'                    header = FALSE, as.is = TRUE) %>%
#'  setNames(c("seqnames", "start", "end", "strand", "motif", "annot",
#'            "uniqreads", "mmreads", "maxoverhang")) %>%
#'  dplyr::mutate(strand = replace(strand, strand == 1, "+")) %>%
#'  dplyr::mutate(strand = replace(strand, strand == 2, "-")) %>%
#'  dplyr::select(seqnames, start, end, strand, uniqreads, mmreads) %>%
#'  dplyr::mutate(seqnames = as.character(seqnames))
#' combCov <- combineCoverages(junctionCounts = jcov,
#'                             junctionPredCovs = txsc$junctionPredCovs,
#'                             txQuants = txsc$txQuants)
#' jcc <- calculateJCCScores(junctionCovs = combCov$junctionCovs,
#'                           geneQuants = combCov$geneQuants)
#' bwFile <- system.file("extdata/reads.chr22.bw", package = "jcc")
#' plotGeneSummary(gtf = gtf, junctionCovs = jcc$junctionCovs,
#'                 useGene = "ENSG00000070371", bwFile = bwFile,
#'                 txQuants = combCov$txQuants)
#' }
#'
#' @importFrom rtracklayer import
#' @importFrom scatterpie geom_scatterpie
#' @importFrom Gviz GeneRegionTrack GenomeAxisTrack DataTrack
#' @importFrom ggplot2 ggplot aes geom_bar theme ylab scale_fill_manual
#'   geom_abline xlab theme_bw expand_limits coord_equal facet_wrap guides
#'   element_text guide_legend
#' @importFrom dplyr filter mutate
#' @importFrom GenomicRanges reduce
#' @importFrom IRanges overlapsAny
#' @importFrom grDevices colorRampPalette dev.off png
#' @importFrom GenomeInfoDb seqnames
#' @importFrom cowplot plot_grid ggdraw draw_image plot_grid get_legend
#' @importFrom S4Vectors %in%
#' @importFrom stats runif
#'
plotGeneSummary <- function(gtf, junctionCovs, useGene, bwFile,
                            txQuants, txCol = "transcript_id",
                            geneCol = "gene_id", exonCol = "exon_id") {
  options(ucscChromosomeNames = FALSE)

  genemodels <- rtracklayer::import(gtf)
  idx <- match(c(txCol, geneCol, exonCol), colnames(mcols(genemodels)))
  colnames(mcols(genemodels))[idx] <- c("transcript", "gene", "exon")
  S4Vectors::mcols(genemodels)$symbol <- mcols(genemodels)$transcript
  genemodels <- subset(genemodels, type == "exon")

  jl <- junctionCovs %>% dplyr::filter(gene == useGene)

  names(bwFile) <- ""
  bwCond <- structure("g1", names = "")

  ## Gene model track
  if (!("gene_name" %in% colnames(S4Vectors::mcols(genemodels)))) {
    S4Vectors::mcols(genemodels)$gene_name <-
      S4Vectors::mcols(genemodels)[[geneCol]]
  }
  gm <- subset(genemodels, tolower(gene) == tolower(useGene) |
                 tolower(gene_name) == tolower(useGene))
  gm <- subset(gm, gene == gene[1])
  id <- unique(gm$gene_name)
  idshow <- paste0(id, " (", unique(gm$gene), ")")
  show_chr <- unique(seqnames(gm))[1]
  gm <- subset(gm, seqnames == show_chr)
  min_coord <- min(start(gm)) - 0.2*(max(end(gm)) - min(start(gm)))
  max_coord <- max(end(gm)) + 0.05*(max(end(gm)) - min(start(gm)))
  gm$transcript <- factor(gm$transcript, levels = unique(gm$transcript))

  ## Other features in the considered region
  gmo <- genemodels[IRanges::overlapsAny(
    genemodels,
    GRanges(seqnames = show_chr,
            ranges = IRanges(start = min_coord,
                             end = max_coord),
            strand = "*"))]
  gmo <- gmo[!(S4Vectors::`%in%`(gmo, gm))]
  gmo <- GenomicRanges::reduce(gmo)

  ## Define colors
  muted <- c("#DC050C","#E8601C","#7BAFDE","#1965B0","#B17BA6",
             "#882E72","#F1932D","#F6C141","#F7EE55","#4EB265",
             "#90C987","#CAEDAB","#777777")
  txs <- levels(gm$transcript)
  ncols <- nlevels(gm$transcript)
  cols <- grDevices::colorRampPalette(muted)(ncols)
  names(cols) <- txs

  grtr <- Gviz::GeneRegionTrack(gm, showId = TRUE, col = NULL,
                                fill = cols[gm$transcript], name = "",
                                background.title = "transparent",
                                col.title = "black", min.height = 15)
  grtr2 <- Gviz::GeneRegionTrack(gmo, showId = TRUE, col = "black",
                                 fill = "white", name = "", col.title = "black",
                                 showId = FALSE, min.height = 15,
                                 background.title = "transparent")

  gtr <- Gviz::GenomeAxisTrack()

  tracks <- c(gtr, grtr, grtr2)

  rnaseqtr1 <- Gviz::DataTrack(range = bwFile,
                               type = "histogram",
                               chromosome = unique(GenomeInfoDb::seqnames(gm)),
                               col.title = "black",
                               fill = "grey",
                               col = "grey",
                               col.histogram = "grey",
                               fill.histogram = "grey",
                               cex.title = 0)
  tracks <- c(rnaseqtr1, tracks)

  rn <- round(1e6*stats::runif(1))
  tmpdir <- tempdir()
  grDevices::png(paste0(tmpdir, "/gviz", rn, ".png"), width = 10.5,
                 height = 5.25, unit = "in", res = 400)
  Gviz::plotTracks(tracks, chromosome = show_chr,
                   from = min_coord, to = max_coord, main = idshow,
                   min.width = 0, min.distance = 0, collapse = FALSE)
  grDevices::dev.off()

  tpms <- ggplot2::ggplot(
    txQuants %>% dplyr::filter(gene == useGene) %>%
      dplyr::mutate(
        transcript = factor(transcript,
                            levels = levels(gm$transcript))),
    ggplot2::aes(x = method, y = TPM, fill = transcript)) +
    ggplot2::geom_bar(stat = "identity", position = "fill") +
    ggplot2::ylab("Relative TPM") + ggplot2::xlab("") +
    ggplot2::theme_bw() +
    ggplot2::scale_fill_manual(values = cols, name = "") +
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, vjust = 0.5,
                                                       hjust = 1, size = 9),
                   legend.text = ggplot2::element_text(size = 7))

  for (tt in txs) {
    jl[[tt]] <- grepl(tt, jl$transcript)
  }
  jl[, txs] <- sweep(jl[, txs], 1, rowSums(jl[, txs]), "/")

  jcov <- ggplot2::ggplot() + ggplot2::geom_abline(intercept = 0, slope = 1) +
    scatterpie::geom_scatterpie(ggplot2::aes(x = uniqreads, y = scaledCov,
                                             r = max(scaledCov)/13),
                                cols = txs, data = jl, color = NA) +
    ggplot2::facet_wrap(~ methodscore, nrow = 2) +
    ggplot2::coord_equal(ratio = 1) +
    ggplot2::expand_limits(x = range(c(0, jl$scaledCov, jl$uniqreads)),
                           y = range(c(0, jl$scaledCov, jl$uniqreads))) +
    ggplot2::scale_fill_manual(values = cols, name = "") +
    ggplot2::xlab("Number of uniquely mapped reads spanning junction") +
    ggplot2::ylab("Scaled predicted junction coverage") +
    ggplot2::theme_bw() +
    ggplot2::theme(strip.text = ggplot2::element_text(size = 10),
                   legend.text = ggplot2::element_text(size = 7))

  plt <- cowplot::plot_grid(
    cowplot::ggdraw() +
      cowplot::draw_image(paste0(tmpdir, "/gviz", rn, ".png")),
    cowplot::plot_grid(tpms + ggplot2::theme(legend.position = "none"),
                       jcov + ggplot2::theme(legend.position = "none"),
                       nrow = 1, labels = c("B", "C"), rel_widths = c(0.9, 1)),
    cowplot::get_legend(
      tpms + ggplot2::theme(legend.direction = "horizontal",
                            legend.justification = "center",
                            legend.box.just = "bottom") +
        ggplot2::guides(fill = ggplot2::guide_legend(nrow =
                                                       length(txs) %/% 8 + 1))),
    ncol = 1, rel_heights = c(1, 0.7, 0.1),
    labels = c("A", "", ""))

  unlink(paste0(tmpdir, "/gviz", rn, ".png"))
  plt
}
